src/targets/gpu/mlir_conv.cpp

```diff
@@ -0,0 +1,214 @@
+
+        auto bin_i = binary_map.find(mlir_options);
+        if (bin_i == binary_map.end()) {
+
+          printf("make_mlir_binary - %s\n", mlir_options.c_str());
+        
+          size_t bin_size = 0;
+          char *bin_buffer = nullptr;
+
+          auto mlir_handle = CreateMlirHandle(mlir_options.c_str());
+
+          if (MlirLowerBin(mlir_handle) == EMlirSuccess &&
+              MlirGenIgemmBin(mlir_handle, &bin_buffer, &bin_size) == EMlirSuccess) {
+            {
+              // static int i=0;
+              // std::string fname = std::string("mlir_conv2d_fwd.") + std::to_string(i++);
+              // auto* F = fopen(fname.c_str(), "w");
+              // fwrite(bin_buffer, 1, bin_size, F);
+              // fclose(F);
+            }
+
+            size_t grid_size, block_size;
+            if (MlirGetExecutionDims(mlir_handle, &grid_size, &block_size) == EMlirSuccess) {
+              printf("make_mlir_binary - grid=%zu block=%zu\n", grid_size, block_size);
+              size_t global_size = grid_size * block_size;
+              result = new execution_spec{{bin_buffer, bin_size}, global_size, block_size};
+            }
+          }
+          DestroyMlirHandle(mlir_handle);
+        
+          binary_map[mlir_options] = result;
+        } else {
+          result = bin_i->second;
+        }
+#endif // MIGRAPHX_MLIR_MIOPEN_SUPPORT
+        return result;
+    }
+
+
+    operation make_code_object_op(instruction_ref op_r, execution_spec *spec) {
+      auto inp_t = op_r->inputs().at(0)->get_shape();
+      auto flt_t = op_r->inputs().at(1)->get_shape();
+      auto out_t = op_r->get_shape();
+      std::vector<shape> expected_inputs = {flt_t, inp_t};
+
+      return migraphx::make_op("gpu::code_object",
+                             {{"code_object", spec->binary},
+                              {"symbol_name", mlir_kernel_name},
+                              {"global", spec->global_size},
+                              {"local", spec->local_size},
+                              {"expected_inputs", migraphx::to_value(expected_inputs)},
+                              {"output", migraphx::to_value(out_t)}});
+    }
+
+    void apply()
+    {
+        init();
+        for(auto it = prog->begin(); it != prog->end(); it++) {
+            if(apply_map.count(it->name()) > 0) {
+                apply_map.at(it->name())(it);
+            }
+        }
+    }
+
+    void add_convolution_op()
+    {
+        apply_map.emplace("convolution", [=](instruction_ref ins) {
+            auto conv_bin = make_mlir_binary(ins);
+            if (conv_bin != nullptr) {
+              auto conv = make_code_object_op(ins, conv_bin);
+              auto inp = ins->inputs().at(0);
+              auto flt = ins->inputs().at(1);
+              return prog->replace_instruction(ins, conv, {flt, inp});
+            }
+            return ins;
+        });
+    }
+
+};
+
+void target_mlir_conv::apply(module& p) const { mlir_apply{&p, this}.apply(); }
```

Comment:
This follow the lowering passes, but I think it makes it more complicated in this case. You could use the matchers:

```cpp
struct find_mlir_convolution
{
    context * ctx = nullptr;
    auto matcher() const
    {
        return match::name("convolution");
    }

    void apply(module& m, match::matcher_result r) const
    {
        auto ins = r.result;
        // Replace ins with gpu::convolution
        ...
    }
};

void target_mlir_conv::apply(module& m) const 
{ 
    find_matches(m, find_mlir_convolution{&ctx});
}
```

Which could extend this to match more operators together(like conv+bias+activation).

However, if you just want to look at convolution, an even simpler way is to just use a simple for loop:

```cpp
void target_mlir_conv::apply(module& m) const 
{ 
    for(auto ins:iterator_for(m))
    {
        if (ins->name() != "convolution")
            continue;
        // Replace ins with gpu::convolution
        ...
    }
}
```